lib/Conversion/FIRRTLToHW/LowerToHW.cpp

```diff
@@ -1468,6 +1492,48 @@ Value FIRRTLLowering::getLoweredValue(Value value) {
   return result;
 }
 
+/// Returns a base type from a vector type.
+static Type getVectorBaseType(Type srcType) {
+  return TypeSwitch<Type, Type>(srcType)
+      .Case<IntegerType>([&](auto intType) { return intType; })
+      .Case<hw::ArrayType>([&](auto vectorType) {
+        return getVectorBaseType(vectorType.getElementType());
+      })
+      .Default([&](auto) { return Type(); });
+}
+
+/// Apply `fn` to each array element and construct new array.
+Value FIRRTLLowering::mapArray(Value array, std::function<Value(Value)> fn) {
+  SmallVector<Value> resultBuffer;
+  bool success = true;
+  std::function<void(Value, Type)> recurse = [&](Value value, Type type) {
+    TypeSwitch<Type>(type)
+        .Case<hw::ArrayType>([&](auto a) {
+          int size = resultBuffer.size();
+          for (size_t i = 0, e = a.getSize(); i != e; ++i) {
+            auto iIdx = getOrCreateIntConstant(log2(e + 1), i);
```

Comment:
Plz don't use log2.

Maybe it would help to add a ArrayGetOp builder that takes a constant value, and creates the type for clients?

Reply:
`FIRRTLLowering` materialize constants through `getOrCreateIntConstant` so I would not use the builder for now.